File: Data-Structures/04_Sorting/03_Quick_Sort.cpp
#include <iostream>
#include <vector>

using std::vector;

void quickSort(vector<int>& arr, int left, int right) {
    if (left >= right) {
        return;
    }

    int pivot = arr[right];
    int i = left; // pointer for the left side (top)

    // Partition: elements smaller than pivot on the left side
    for (int j = left; j < right; j++) { // pointer for the right side (bottom)
        if (arr[j] < pivot) {
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
            i++;
        }
    }

    // Move pivot in-between left and right sides
    arr[right] = arr[i];
    arr[i] = pivot;

    // Quick sort left and right sides using recursion
    quickSort(arr, left, i - 1);
    quickSort(arr, i + 1, right);
}

int main() {
    vector<int> arr = {6, 2, 4, 1, 3};
    int left = 0;
    int right = arr.size() - 1;

    quickSort(arr, left, right);

    for (int i = 0; i < arr.size(); i++) {
        printf("%d, ", arr[i]);
    }
    printf("\n");

    return 0;
}

File: DSA-for-Beginners/04_Sorting/03_Quick_Sort.py
def quickSort(arr, left, right):
    if left >= right:
        return arr
    
    pivot = arr[right]
    i = left # pointer for left side

    # Partition: elements smaller than pivot on left side
    for j in range(left, right):
        if arr[j] < pivot:
            tmp = arr[i]
            arr[i] = arr[j]
            arr[j] = tmp
            i += 1
    
    # Move pivot in-between left and right sides
    arr[right] = arr[i]
    arr[i] = pivot

    # Quick sort left and right side
    quickSort(arr, left, i - 1)
    quickSort(arr, i + 1, right)

    return arr

def main():
    arr = [6, 2, 4, 1, 3]
    left = 0
    right = len(arr) - 1

    quickSort(arr, left, right)
    print(arr)

if __name__ == "__main__":
    main()